### ISSUE TITLE
GodotPhysics 3D crash with multi-threaded physics

### ISSUE THREAD

**Issue by akien-mga** [godotengine/godot#90689]
### Tested versions

- Reproducible in 4.2.1-stable
- Reproducible in 4.3.dev (c951421c996467dcb7106a33774410a7b5258690, after #90268) but also 4.3.dev (a44b0b6dc1ccd6dc364bb72c426e844d6c85744e, before #90268), so it's not a regression, but that PR also didn't fix it

### System information

Fedora Linux 39 (KDE Plasma) - Wayland - Vulkan (Forward+) - dedicated AMD Radeon RX 7600M XT (RADV NAVI33) () - AMD Ryzen 7 7840HS w/ Radeon 780M Graphics (16 Threads)

### Issue description

When enabling Multi-threaded physics in the [GDQuest TPS demo], I easily trigger this crash on Linux just by shooting a few times (hold right click to aim, left click to shoot).

It's not necessarily _only_ linked to shooting though, I tried to isolate the player scene in a new scene with just a StaticBody3D floor to stand on, and there shooting doesn't trigger a crash. So something else in the main scene combined with shooting triggers the issue.

While I got it to crash by just shooting in the sky, shooting a stack of crates is a sure-fire way to reproduce the bug.

Backtrace:
```
ERROR: FATAL: Index p_index = 0 is out of bounds (shapes.size() = 0).
   at: get_shape_transform (./servers/physics_3d/godot_collision_object_3d.h:128)

Thread 12 "godot-git" received signal SIGILL, Illegal instruction.
[Switching to Thread 0x7fffe38006c0 (LWP 31046)]
0x000000000715c298 in GodotCollisionObject3D::get_shape_transform (this=0x11512710, p_index=0) at ./servers/physics_3d/godot_collision_object_3d.h:128

(gdb) bt
#0  0x000000000715c298 in GodotCollisionObject3D::get_shape_transform (this=0x11512710, p_index=0) at ./servers/physics_3d/godot_collision_object_3d.h:128
#1  0x0000000007110e66 in GodotArea2Pair3D::setup (this=0x7fff98af0720, p_step=0.0166666675) at ./servers/physics_3d/godot_area_pair_3d.cpp:126
#2  0x0000000007144b85 in GodotStep3D::_setup_constraint (this=0x7fff98000b80, p_constraint_index=54, p_userdata=0x0) at ./servers/physics_3d/godot_step_3d.cpp:117
#3  0x000000000719968b in WorkerThreadPool::GroupUserData<GodotStep3D, void (GodotStep3D::*)(unsigned int, void*), decltype(nullptr)>::callback_indexed(unsigned int) (this=0x7fff9818aa20, p_index=54)
    at ./core/object/worker_thread_pool.h:184
#4  0x0000000007a683af in WorkerThreadPool::_process_task (this=0xb4b0200, p_task=0xdd0e270) at ./core/object/worker_thread_pool.cpp:92
#5  0x0000000007a68b3a in WorkerThreadPool::_thread_function (p_user=0xb562aa0) at ./core/object/worker_thread_pool.cpp:193
#6  0x00000000074bb98a in Thread::callback (p_caller_id=12, p_settings=..., p_callback=0x7a6899a <WorkerThreadPool::_thread_function(void*)>, p_userdata=0xb562aa0) at ./core/os/thread.cpp:64
#7  0x00000000074d14a7 in std::__invoke_impl<void, void (*)(unsigned long, Thread::Settings const&, void (*)(void*), void*), unsigned long, Thread::Settings, void (*)(void*), void*> (
    __f=@0xb560a08: 0x74bb8fc <Thread::callback(unsigned long, Thread::Settings const&, void (*)(void*), void*)>) at /usr/include/c++/13/bits/invoke.h:61
#8  0x00000000074d1370 in std::__invoke<void (*)(unsigned long, Thread::Settings const&, void (*)(void*), void*), unsigned long, Thread::Settings, void (*)(void*), void*> (
    __fn=@0xb560a08: 0x74bb8fc <Thread::callback(unsigned long, Thread::Settings const&, void (*)(void*), void*)>) at /usr/include/c++/13/bits/invoke.h:96
#9  0x00000000074d122b in std::thread::_Invoker<std::tuple<void (*)(unsigned long, Thread::Settings const&, void (*)(void*), void*), unsigned long, Thread::Settings, void (*)(void*), void*> >::_M_invoke<0ul, 1ul, 2ul, 3ul, 4ul> (this=0xb5609e8) at /usr/include/c++/13/bits/std_thread.h:292
#10 0x00000000074cfe1e in std::thread::_Invoker<std::tuple<void (*)(unsigned long, Thread::Settings const&, void (*)(void*), void*), unsigned long, Thread::Settings, void (*)(void*), void*> >::operator() (
    this=0xb5609e8) at /usr/include/c++/13/bits/std_thread.h:299
#11 0x00000000074cee56 in std::thread::_State_impl<std::thread::_Invoker<std::tuple<void (*)(unsigned long, Thread::Settings const&, void (*)(void*), void*), unsigned long, Thread::Settings, void (*)(void*), void*> > >::_M_run (this=0xb5609e0) at /usr/include/c++/13/bits/std_thread.h:244
#12 0x0000000007f88533 in execute_native_thread_routine ()
#13 0x00007ffff7d6d897 in start_thread () from /lib64/libc.so.6
#14 0x00007ffff7df4a5c in clone3 () from /lib64/libc.so.6
```

CC @RandomShaper @godotengine/physics

*Edit:* I tested with godot-jolt 0.12.0, but it doesn't support multi-threaded physics currently, so naturally the bug isn't reproducible. So we can't assess yet whether the bug is specific to GodotPhysics or to the MT code.

### Steps to reproduce

- Download GDQuest TPS demo
- Apply this diff to enable multi-threaded physics
```diff
diff --git a/project.godot b/project.godot
index 33dfc57..64b0cd6 100644
--- a/project.godot
+++ b/project.godot
@@ -114,4 +114,5 @@ pause={
 
 [physics]
 
+3d/run_on_separate_thread=true
 3d/default_gravity=16.0
```

### Minimal reproduction project (MRP)

Not minimal, but for this kind of bug an open source project like this should be suitable: https://github.com/gdquest-demos/godot-4-3d-third-person-controller

**Comment by akien-mga** [comment]
Might be a duplicate of #70491, I'll test.

**Comment by akien-mga** [comment]
This is fixed by #72491 like the MRP in #70491, and the backtraces are similar, so this is indeed a duplicate. Consolidating there.
